src/Microsoft.AspNet.Security.OpenIDConnect/OpenidConnectAuthenticationHandler.cs

```diff
@@ -0,0 +1,621 @@
+                        RedirectUri = ticket.Properties.Dictionary.ContainsKey(OpenIdConnectAuthenticationDefaults.RedirectUriUsedForCodeKey) ?
+                                      ticket.Properties.Dictionary[OpenIdConnectAuthenticationDefaults.RedirectUriUsedForCodeKey] : string.Empty,
+                    };
+
+                    await Options.Notifications.AuthorizationCodeReceived(authorizationCodeReceivedNotification);
+                    if (authorizationCodeReceivedNotification.HandledResponse)
+                    {
+                        return authorizationCodeReceivedNotification.AuthenticationTicket;
+                    }
+
+                    if (authorizationCodeReceivedNotification.Skipped)
+                    {
+                        return null;
+                    }
+                }
+
+                return ticket;
+            }
+            catch (Exception exception)
+            {
+                // We can't await inside a catch block, capture and handle outside.
+                authFailedEx = ExceptionDispatchInfo.Capture(exception);
+            }
+
+            if (authFailedEx != null)
+            {
+                _logger.WriteError("Exception occurred while processing message: '" + authFailedEx.ToString());
+
+                // Refresh the configuration for exceptions that may be caused by key rollovers. The user can also request a refresh in the notification.
+                if (Options.RefreshOnIssuerKeyNotFound && authFailedEx.SourceException.GetType().Equals(typeof(SecurityTokenSignatureKeyNotFoundException)))
+                {
+                    Options.ConfigurationManager.RequestRefresh();
+                }
+
+                var authenticationFailedNotification = new AuthenticationFailedNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>(Context, Options)
+                {
+                    ProtocolMessage = openIdConnectMessage,
+                    Exception = authFailedEx.SourceException
+                };
+
+                await Options.Notifications.AuthenticationFailed(authenticationFailedNotification);
+                if (authenticationFailedNotification.HandledResponse)
+                {
+                    return authenticationFailedNotification.AuthenticationTicket;
+                }
+
+                if (authenticationFailedNotification.Skipped)
+                {
+                    return null;
+                }
+
+                authFailedEx.Throw();
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// 'Remembers' the nonce associated with this message. By default the nonce added as a secure cookie.
+        /// </summary>
+        /// <param name="message"><see cref="OpenIdConnectMessage"/> associatied with the nonce.</param>
+        /// <remarks>A cookie is added with the name obtained from  <see cref="GetNonceKey"/>.</remarks>
+        public virtual void AddNonce(string nonce)
```

Comment:
We discussed removing all these virtual methods and moving all the extensibility to Notifications, no?